ReSharper.FSharp/src/FSharp.Psi.Features/src/Daemon/Stages/InferredTypeHintStage.fs

```diff
@@ -0,0 +1,175 @@
+            let prefix = variableNameParts.[0].ToLowerInvariant()
+            prefix = "has" || prefix = "is"
+        else
+            false
+
+    let isEvidentFromVariableName (fsType: FSharpType) variableName =
+        if not highlightingContext.HideTypeNameHintsWhenTypeNameIsEvidentFromVariableName then false else
+
+        let typ = fsType.MapType(Array.empty, fsFile.GetPsiModule())
+        if not (typ.IsValid()) then false else
+
+        let variableNameParts = NamesHelper.GetParts(nameParser, namingPolicyProvider, variableName)
+        if isTypeEvidentFromVariableNamePrefix typ variableNameParts then true else
+
+        match typ.GetTypeElement() with
+        | null -> false
+        | typeElement ->
+
+        let typeName = typeElement.ShortName
+        String.Equals(typeName, variableName, StringComparison.OrdinalIgnoreCase) ||
+        not (String.IsNullOrEmpty(typeName)) &&
+        not (String.IsNullOrEmpty(variableName)) &&
+        NamesHelper.IsLike(variableNameParts, NamesHelper.GetParts(nameParser, namingPolicyProvider, typeName))
+
+    let isTypeOfPatternEvident (pattern: ISynPat) =
+
+        // v-- not evident
+        // x::y::z
+        //    ^--^-- evident
+        let tuplePat = TuplePatNavigator.GetByPattern pattern
+        if isNotNull tuplePat then
+            let consPat = ConsPatNavigator.GetByPattern1 tuplePat
+            if isNull consPat then false else
+
+            // Are we nested inside another ConsPat?
+            let parentTuplePat = TuplePatNavigator.GetByPattern consPat
+            if isNull parentTuplePat then tuplePat.Patterns.IndexOf pattern <> 0 else
+            isNotNull (ConsPatNavigator.GetByPattern1 parentTuplePat)
+
+        else
+
+        //  v-- not evident
+        // [x; y; z]
+        //     ^--^-- evident
+        let listOrListPat = ArrayOrListPatNavigator.GetByPattern pattern
+        if isNotNull listOrListPat then
+            listOrListPat.Patterns.IndexOf pattern <> 0
+
+        else
+
+        false
+
+    override x.VisitNode(node, context) =
+        for child in node.Children() do
+            match child with
+            | :? IFSharpTreeNode as treeNode -> treeNode.Accept(x, context)
+            | _ -> ()
+
+    override x.VisitLocalReferencePat(localRefPat, consumer) =
+        let pat = localRefPat.IgnoreParentParens()
+        if isNotNull (TypedPatNavigator.GetByPattern(pat)) then () else
+
+        let binding = BindingNavigator.GetByHeadPattern(pat)
+        if isNotNull binding && isNotNull binding.ReturnTypeInfo then () else
+
+        if isTypeOfPatternEvident pat then () else
+
+        match box (localRefPat.GetFSharpSymbolUse()) with
+        | null -> ()
+        | symbolUse ->
+
+        let symbolUse = symbolUse :?> FSharpSymbolUse
+        match symbolUse.Symbol with
+        | :? FSharpMemberOrFunctionOrValue as mfv when not (isEvidentFromVariableName mfv.FullType localRefPat.ReferenceName.Identifier.Name) ->
```

Comment:
`IReferenceName.ShortName` would be better to use, since it'd check `Identifier` for null first.
A `shortName <> SharedImplUtil.MISSING_DECLARATION_NAME` check before getting the symbol could probably make it a bit better too.